viewing_party/party.py

```diff
@@ -1,23 +1,175 @@
+            friends_watched_titles.append(friend_movie['title'])
+    for movie in watched_movie:
+        if movie['title'] not in friends_watched_titles:
+            user_unique_watched_movies.append(movie)    
+    print(user_unique_watched_movies)
+    return user_unique_watched_movies
+
+
+def get_friends_unique_watched(user_data):
+    friends_unique_watched_movies_list = []
+
+    user_watched_title_set = set()   
+    user_watched_movie_list = user_data.get('watched', [])
+
+    friends_watched_movie_list = user_data.get('friends', [])
+    friends_watched_title_set = set()
+    
+    for user_watched_movies in user_watched_movie_list:
+        user_watched_title_set.add(user_watched_movies['title'])
+
+    for friends_movies_list in friends_watched_movie_list: #list
+        for firends_watched_movie in friends_movies_list.get('watched', []): #dict.value
+            if firends_watched_movie['title'] not in user_watched_title_set and firends_watched_movie['title'] not in friends_watched_title_set:
+                friends_unique_watched_movies_list.append(firends_watched_movie)
+            friends_watched_title_set.add(firends_watched_movie['title'])
+    
+    return friends_unique_watched_movies_list
 
         
-# -----------------------------------------
 # ------------- WAVE 4 --------------------
-# -----------------------------------------
 
-# -----------------------------------------
+def get_available_recs(user_data):
+
+    result = []
+    
+    user_subscriptions = set(user_data.get('subscriptions', []))
+    friends_recommendations = get_friends_unique_watched(user_data)
+
+    for movie in friends_recommendations:
+        if movie['host'] in user_subscriptions:
+            result.append(movie)
+    return result
+
+
 # ------------- WAVE 5 --------------------
-# -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    recommended_movies_list = []
+
+    user_most_frequent_genre = get_most_watched_genre(user_data)
+    friends_unique_watched_movies_list = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_watched_movies_list:
+        if movie['genre'] == user_most_frequent_genre:
+            recommended_movies_list.append(movie)
+    
+    return recommended_movies_list
+
+
+def get_rec_from_favorites(user_data):
+    recommend_movies_list = []
+
+    user_favorites_watched_movies_list = user_data.get('favorites', [])
+
+    friends_watched_movies_list = user_data.get('friends',[])
+   
+    friends_watched_movies_title_list = []
+    for friend_watched_movies in friends_watched_movies_list:
+        for friend_movie in friend_watched_movies.get('watched',[]):
+            friends_watched_movies_title_list.append(friend_movie['title'])
+    for movie in user_favorites_watched_movies_list:
+        if movie['title'] not in friends_watched_movies_title_list:
+            recommend_movies_list.append(movie)
+ 
+    return recommend_movies_list

```

Comment:
Thinking about the requirements of if a movie can be added to the `recommend_movies_list`:

- Determine a list of recommended movies. A movie should be added to this list if and only if:
  - The movie is in the user's "favorites"
  - None of the user's friends have watched it

Don't we have a function that takes care of the second requirement? How could we change our function here so that it utilizes that function?